docs/cardano-monads.rst

```diff
@@ -0,0 +1,773 @@
+        type SlottingContext Prod = Either SimpleSlottingVar NtpSlottingVar
+        getCurrentSlot = do
+          sctx <- asks slottingContext
+          either getCurrentSlotSimple getCurrentSlotNtp sctx
+
+While effect context sums are closed, we are fine with it because we don't
+need extensibility at instance definition site, while the implementations
+themselves are defined independently.
+
+In order to address the issue of extensibility, for each ``Has``-class we can
+define two instances in a generic way::
+
+    class HasFooContext b ctx where
+        fooContext :: ctx -> FooContext b
+
+    instance HasFooContext b (FooContext b, _) where
+        fooContext = fst
+
+    instance {-# OVERLAPPABLE #-} HasFooContext b ctx => HasFooContext b (_, ctx) where
+        fooContext = fooContext . snd
+
+This way it's possible to add a context using ``withReaderT (fooCtx,)``. A
+stylistic issue with this is the use of ``{-# OVERLAPPABLE #-}``, but it seems
+rather benign here. A technical issue is that need ``FooContext`` to be a data
+family (rather than a type family) in order for this to work, but it's a simple
+change.
+
+Let us have a point-by-point rundown of the approach properties:
+
+* Flexibility: GOOD. We can use an effect locally by extending the reader
+  context, provided that the base monad supports it. We don't need to use monad
+  transformers to add an effect. We can run effects partially (by supplying only
+  a part of the context). We can have different prod/test base monads.
+
+* Extensibility: MODERATE. To define an effect we need two classes and a type/data
+  family, and the ``Has``-class presents a certain amount of boilerplate. Other than
+  that, it is good: we don't have the ``N*M`` problem.
+
+* Ease of use: GOOD (hopefully). We don't have lifting, hoisting, etc, just write
+  the code in ``ReaderT`` over an abstract base monad.
+
+* Compile-time performance: GOOD. While we do have type families, they only require
+  a single step of reduction, so this is not an issue.
+
+* Run-time performance: GOOD. Since we use classes, the methods should be
+  specialized to concrete monads, and we pass the context in a flat ``ReaderT``.
+
+* Run-time configurability: GOOD. We identify two categories of effect impls --
+  those that depend on the base monad and those that don't. In the first case
+  we don't need configurability, in the second we use effect context sums.
+
+* Predictability: SUPERB. While the types do affect the choice of implementation,
+  all code is written in a single monad, so it's easy to track which
+  implementation is used where. (We forbid ``IdentityT`` redirection layers).
+  It's extremely easy to find out what code is executed for a particular impl,
+  because it's in a top-level instance, not in a method dictionary.
+
+As you can see, the hybrid approach is a compromise. There are still issues,
+but there are no fatal flaws (we don't score LOW for any of the criteria).
```

Comment:
I've read about the hybrid approach, but it's not obvious how our current code will be transformed if we choose it.
For instance, we have:
```
usNormalize :: (USLocalLogicMode ctx m) => m ()
usNormalize = do
    tip <- DB.getTip
    stateVar <- mvState <$> views (lensOf @UpdateContext) ucMemState
    atomically . writeTVar stateVar =<< usNormalizeDo (Just tip) Nothing
```

What should be the type of this function in new approach?
Second, we have several execution modes: `RealMode`, `AuxxMode`, `ExplorerMode`, etc. Will we have different «base monads» for each mode? Will we have different contexts? How will it look?

Reply:
@int-index please add examples

Reply:
> What should be the type of this function in new approach?

```
usNormalize :: USLocalLogicMode ctx b => ReaderT ctx b ()
```

> Second, we have several execution modes: RealMode, AuxxMode, ExplorerMode, etc. Will we have different «base monads» for each mode?

Not necessarily. For example, `AuxxMode` is the same as `RealMode`, both will run in `Prod` (but `AuxxMode` might have a bigger `ReaderT` context on top).

Not sure about `ExplorerMode`, I tried to find it in our code right now to analyze, but only found `ExplorerTestMode`.

The general rule is that we can reuse a mode as long as we don't need to change its fundamental behavior, and we only need to add more context. But it's okay to define a separate mode for a fundamental change in behavior.

Reply:
> please add examples

To clarify: there are examples how to define new interface (e. g. `MonadFoo` is an interface), how to implement an interface (e. g. `instance MonadDBGet Prod where` is an implementation), but there is no example how to use this interface.
This approach is something new, so having more examples would be cool. Probably also a more real-life example of using `withReaderT (fooCtx,)`.

Reply:
> Not necessarily. For example, AuxxMode is the same as RealMode, both will run in Prod (but AuxxMode might have a bigger ReaderT context on top).

Ok, but for auxx we want to use different logging method, don't we? AFAIU, currently it's solved by passing `PrintAction` explicitly as an argument (which is called `Dictionary-Passing Style` in this document). But we want to use only one approach. Does this hybrid approach allow us to solve this problem neatly?

Reply:
> Does this hybrid approach allow us to solve this problem neatly?

Particularly in case of logging, a special feature was added to `log-warper` to support this, so it's not an issue. But before this feature, I would have to answer that this would require either (1) a separate mode, or (2) an effect context sum.